hibernate-core/src/main/java/org/hibernate/type/descriptor/java/JdbcDateJavaTypeDescriptor.java

```diff
@@ -122,56 +120,81 @@ public int extractHashCode(Date value) {
 		return hashCode;
 	}
 
-	@SuppressWarnings({ "unchecked" })
 	@Override
-	public <X> X unwrap(Date value, Class<X> type, WrapperOptions options) {
+	public Date coerce(Object value, CoercionContext coercionContext) {
+		return wrap( value, null );
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public Object unwrap(Date value, Class type, WrapperOptions options) {
 		if ( value == null ) {
 			return null;
 		}
+
+		if ( LocalDate.class.isAssignableFrom( type ) ) {
+			return unwrapLocalDate( value );
+		}
+
 		if ( java.sql.Date.class.isAssignableFrom( type ) ) {
-			final java.sql.Date rtn = value instanceof java.sql.Date
-					? ( java.sql.Date ) value
-					: new java.sql.Date( value.getTime() );
-			return (X) rtn;
+			return unwrapSqlDate( value );
 		}
-		if ( java.sql.Time.class.isAssignableFrom( type ) ) {
-			final java.sql.Time rtn = value instanceof java.sql.Time
-					? ( java.sql.Time ) value
-					: new java.sql.Time( value.getTime() );
-			return (X) rtn;
+
+		if ( java.util.Date.class.isAssignableFrom( type ) ) {
+			return value;
 		}
-		if ( java.sql.Timestamp.class.isAssignableFrom( type ) ) {
-			final java.sql.Timestamp rtn = value instanceof java.sql.Timestamp
-					? ( java.sql.Timestamp ) value
-					: new java.sql.Timestamp( value.getTime() );
-			return (X) rtn;
+
+		if ( Long.class.isAssignableFrom( type ) ) {
+			return unwrapDateEpoch( value );
 		}
-		if ( Date.class.isAssignableFrom( type ) ) {
-			return (X) value;
+
+		if ( String.class.isAssignableFrom( type ) ) {
+			return toString( value );
 		}
+
 		if ( Calendar.class.isAssignableFrom( type ) ) {
 			final GregorianCalendar cal = new GregorianCalendar();
-			cal.setTimeInMillis( value.getTime() );
-			return (X) cal;
+			cal.setTimeInMillis( unwrapDateEpoch( value ) );
+			return cal;
 		}
-		if ( Long.class.isAssignableFrom( type ) ) {
-			return (X) Long.valueOf( value.getTime() );
+
+
+		if ( java.sql.Time.class.isAssignableFrom( type ) ) {
+			throw new IllegalArgumentException( "Illegal attempt to treat `java.sql.Date` as `java.sql.Time`" );
 		}
-		if ( LocalDate.class.isAssignableFrom( type ) ) {
-			if ( value instanceof java.sql.Date ) {
-				return (X) ( (java.sql.Date) value ).toLocalDate();
-			}
+
+		if ( java.sql.Timestamp.class.isAssignableFrom( type ) ) {
+			throw new IllegalArgumentException( "Illegal attempt to treat `java.sql.Date` as `java.sql.Timestamp`" );
```

Comment:
I don't mind the change, but is it really illegal?

Reply:
I wondered this myself.  Its a bit non-sense, but not not doable.